Добавлена поддержка traits и interfaces

avasilenko34 opened the pull request.

Commit:
Добавлена поддержка traits и interfaces

webarchitect609 commented on src/main/Taxidermist.php:
```diff
@@ -126,7 +126,7 @@ public function taxidermize(string $bitrixClass): ?bool
             return false;
```
Выше есть ещё одна проверка `class_exists()`. Надо разобраться, для чего она нужна, и не нуждается ли она тоже в исправлении.

webarchitect609 commented on src/main/Taxidermist.php:
```diff
@@ -126,7 +126,7 @@ public function taxidermize(string $bitrixClass): ?bool
         }
         $mockClass = self::MOCK_NAMESPACE . ltrim($bitrixClass, '\\');
-        if (!class_exists($mockClass)) {
+        if (!class_exists($mockClass) && !interface_exists($mockClass) && !trait_exists($mockClass)) {
```
Нужен Unit-тест, который бы повторял ошибку, и гарантировал, что она не повторяется больше.

avasilenko34 committed:
Merge pull request #1 from avasilenko34/feature/add_cache_interfaces

Feature/add cache interfaces

Commit:
Merge branch 'master' into feature/add-support-for-interfaces-and-traits

Commit:
feat: Добавлены классы:
Bitrix\Main\Error
Bitrix\Main\ErrorCollection
Bitrix\Main\Type\Dictionary

Добавлен трейт:
Bitrix\Main\ErrorableImplementation

Commit:
fix: Исправлена работа с интерфейсами и трейтами

Commit:
test: Добавлены unit-тесты
testTaxidermizeInterface
testTaxidermizeTrait

webarchitect609 commented on src/test/TaxidermistTest.php:
```diff
@@ -96,6 +96,41 @@ public function testTaxidermizeAll()
     }
 
+    /**
+     * @depends testTaxidermize
```
Зачем здесь зависимость от `testTaxidermize`? Разве без неё не работает?

webarchitect609 replied:
Зависимость `@depends testTaxidermize` следует удалить.

webarchitect609 commented on src/test/TaxidermistTest.php:
```diff
@@ -96,6 +96,41 @@ public function testTaxidermizeAll()
+    }
+
+    /**
+     * @depends testTaxidermize
```
Аналогично. Зачем здесь зависимость от `testTaxidermize`? Разве без неё не работает?

webarchitect609 replied:
Аналогично. Зависимость `@depends testTaxidermize` следует удалить.

webarchitect609 commented on src/test/TaxidermistTest.php:
```diff
@@ -96,6 +96,41 @@ public function testTaxidermizeAll()
+    {
+        $traitName = 'Bitrix\Main\ErrorableImplementation';
+
+        // Внутренняя зависимость
```
А где эта внутренняя зависимость? Не вижу, чтобы `Bitrix\Main\Type\Dictionary` использовал `Bitrix\Main\ErrorableImplementation`.

webarchitect609 replied:
Если по факту зависимости нет, следует удалить лишний код таксидермации класса `Bitrix\Main\Type\Dictionary`.

webarchitect609 commented on src/main/Mock/Bitrix/Main/Error.php:
```diff
@@ -0,0 +1,76 @@
+     * which is a value of any type other than a resource.
+     * @since 5.4.0
+     */
+    public function jsonSerialize()
```
Предлагаю вернуть тело этому методу. Вдруг, в будущем понадобится в каком-нибудь Unit-тесте проверять правильность или неправильность сериализации в JSON этого объекта.

webarchitect609 commented on src/main/Mock/Bitrix/Main/ErrorCollection.php:
```diff
@@ -0,0 +1,60 @@
+    /**
+     * Adds an error to the collection.
+     * @param Error $error An error object.
+     * @param $offset Offset in the array.
```
Тут явная ошибка, которую надо исправить. PhpStorm ругается `Undefined class 'Offset' `. По исходному коду Битрикс надо определить, какой там на самом деле тип данных, и подставить его в phpDoc.

Т.е. можно обобщить рекомендацию так: сам тип аргумента у метода менять не следует, а вот phpDoc-блоки следует исправлять так, чтобы PhpStorm на них не ругался, и чтобы в них была информация, добытая из исходного кода Битрикс о реальном типе используемых данных.

webarchitect609 reviewed (changes requested):
Также следует:
- пересмотреть ещё раз все новые классы/трейты/интерфейсы, чтобы исправить там всё, чтобы PhpStorm не ругался;
- аккуратно внести изменения в `CHANGELOG.md` файл.

Commit:
feat: Добавлены классы:
Bitrix\Main\Error
Bitrix\Main\ErrorCollection
Bitrix\Main\Type\Dictionary
Добавлен трейт:
Bitrix\Main\ErrorableImplementation

fix: Исправлена работа с интерфейсами и трейтами

test: Добавлены unit-тесты
testTaxidermizeInterface
testTaxidermizeTrait

webarchitect609 commented on src/main/Mock/Bitrix/Main/ErrorCollection.php:
```diff
@@ -41,7 +43,7 @@ public function getErrorByCode($code)
      * Adds an error to the collection.
      * @param Error $error An error object.
-     * @param $offset Offset in the array.
+     * @param $offset. Offset in the array.
```
Ну вот зачем ты так сделал? Я же просил проверить исходный код, чтобы выяснить какой(или какие) типы данных на самом деле здесь используются.

avasilenko34 replied:
Это именно ошибка в phpDoc со стороны битрикса. Класс Offset - не существует. А если посмотреть на исходники, то там видно, что на вход явно ожидается не объект.

webarchitect609 commented on src/test/TaxidermistTest.php:
```diff
@@ -96,6 +96,36 @@ public function testTaxidermizeAll()
+    /**
+     * @return void
+     */
+    public function testTaxidermizeInterface()
```
Тест неверный! Его самостоятельный запуск невозможен: он всегда падает с ошибкой! Так быть не должно. Да и с чего возвращать `false`, если установка алиаса на интерфейс прошла успешно. Надо разобраться и поправить.

Commit:
fix: Исправлен phpDoc для метода ErrorCollection::setError
Исправлены тесты testTaxidermizeInterface и testTaxidermizeTrait

Commit:
fix: Исправление версии в CHANGELOG.md

Commit:
Merge branch 'master' of https://github.com/webarchitect609/bitrix-taxidermist into feature/add-support-for-interfaces-and-traits
